addons/delivery/models/delivery_carrier.py

```diff
@@ -123,3 +136,151 @@ def cancel_shipment(self, pickings):
+
+            if record.delivery_type == 'fixed':
+                lines = PriceRule.search([('carrier_id', '=', record.id)])
+                if lines:
+                    lines.unlink()
+
+            #create the delivery price rules
+            if record.free_if_more_than:
+                line_data = {
+                    'carrier_id': record.id,
+                    'name': _('Free if more than %.2f') % record.amount,
+                    'variable': 'price',
+                    'operator': '>=',
+                    'max_value': record.amount,
+                    'standard_price': 0.0,
+                    'list_price': 0.0,
+                }
+                PriceRule.create(line_data)
+            if isinstance(record.fixed_price, float):
+                line_data = {
+                    'carrier_id': record.id,
+                    'name': _('Default price'),
+                    'variable': 'price',
+                    'operator': '>=',
+                    'max_value': 0.0,
+                    'standard_price': record.fixed_price,
+                    'list_price': record.fixed_price,
+                }
+                PriceRule.create(line_data)
+        return True
+
+    @api.model
+    def create_product_for_delivery(self, vals):
+        if not vals.get('product_id'):
+            value = {
+                'name': '%s delivery' % (vals['name']),
+                'default_code': 'Delivery',
+                'type': 'service',
+                'standard_price': 10.0,
+                'list_price': 10.0
+            }
+            vals['product_id'] = self.env['product.product'].create(value).id
+        return vals
+
+    @api.model
+    def filter_state_country(self, vals):
+        if vals.get('country_ids'):
+            country_ids = [country['id'] for country in self.resolve_2many_commands('country_ids', vals.get('country_ids'))]
+            state_ids = [state['id'] for state in self.resolve_2many_commands('state_ids', vals.get('state_ids'))]
+            vals['state_ids'] = [(6, 0, self.env['res.country.state'].browse(state_ids).filtered(lambda x: x.country_id.id in country_ids).ids)]
+        return vals
+
+    @api.multi
+    def write(self, vals):
+        if vals.get('country_ids'):
+            country_ids = [country['id'] for country in self.resolve_2many_commands('country_ids', vals.get('country_ids'))]
+            for carrier in self:
+                if carrier.state_ids:
+                    vals['state_ids'] = [(3, state.id) for state in carrier.state_ids.filtered(lambda x: x.country_id.id not in country_ids)]
+        res = super(DeliveryCarrier, self).write(vals)
+        self.create_price_rules(vals)
+        return res
+
+    @api.model
+    def create(self, vals):
+        vals = self.filter_state_country(vals)
+        vals = self.create_product_for_delivery(vals)
+        res = super(DeliveryCarrier, self).create(vals)
+        res.create_price_rules(vals)
+        return res
+
+    @api.multi
```

Comment:
hello @atp-odoo, If i can use api.one then it will return list and i need to use list with [0] structure so use api.multi with ensure_one  Thanks